src/Microsoft.ML.TimeSeries/SrCnnEntireTransformer.cs

```diff
@@ -0,0 +1,905 @@
+            Host.Assert(OutputLength == AnomalyOnlyOutputLength || OutputLength == AnomalyAndExpectedValueOutputLength || OutputLength == AnomalyAndMarginOutputLength);
+
+            // *** Binary format ***
+            // <base>
+
+            // for each added column
+            //   double: threshold
+            //   int: batch size
+            //   int: detect mode
+            //   double: sensitivity
+            //   int: output length
+            //   int: dictionary size
+            //   long stream: dictionary keys
+            //   double stream: dictionary values
+
+            base.SaveColumns(ctx);
+            ctx.Writer.Write(_options.Threshold);
+            ctx.Writer.Write(_options.BatchSize);
+            ctx.Writer.Write((int)_options.SrCnnDetectMode);
+            ctx.Writer.Write(_options.Sensitivity);
+            ctx.Writer.Write(OutputLength);
+
+            ctx.Writer.Write(_anomalyDict.Count);
+
+            List<long> keyList = new List<long>();
+            List<double> valList = new List<double>();
+            foreach (var pair in _anomalyDict)
+            {
+                keyList.Add(pair.Key);
+                valList.AddRange(pair.Value);
+            }
+            ctx.Writer.WriteLongStream(keyList);
+            ctx.Writer.WriteDoubleStream(valList);
+        }
+
+        private protected override IRowMapper MakeRowMapper(DataViewSchema schema) => new Mapper(this, schema);
+
+        private sealed class Mapper : MapperBase
+        {
+            private readonly SrCnnEntireTransformer _parent;
+            private readonly VBuffer<ReadOnlyMemory<char>> _slotNames;
+            private readonly int _timestampColumnIndex;
+            private static readonly string[] _outputFields = new string[] { "Is Anomaly", "Anomaly Score", "Mag", "Expected Value", "Boundary Unit", "Upper Boundary", "Lower Boundary" };
+
+            public Mapper(SrCnnEntireTransformer parent, DataViewSchema inputSchema)
+                : base(parent.Host.Register(nameof(Mapper)), inputSchema, parent)
+            {
+                Contracts.AssertValue(parent);
+                _parent = parent;
+
+                SourceColumn inputCol = SourceColumn.Parse(_parent.ColumnPairs[0].inputColumnName);
+
+                if (!inputSchema.TryGetColumnIndex(inputCol.Source[0], out _timestampColumnIndex))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "inputTimestamp", inputCol.Source[0]);
+                if (!(inputSchema[_timestampColumnIndex].Type is DateTimeDataViewType))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "inputTimestamp", inputCol.Source[0], "DateTime", inputSchema[_timestampColumnIndex].Type.ToString());
+
+                _slotNames = new VBuffer<ReadOnlyMemory<char>>(_parent.OutputLength, _outputFields.ToList().GetRange(0, _parent.OutputLength).Select(x => x.AsMemory()).ToArray());
```

Comment:
>ToList().GetRange(0, _parent.OutputLength) [](start = 99, length = 42)

ToList().GetRange() may not be necessary. You can call Take() on the array

Reply:
Fixed.

Reply:
> > ToList().GetRange(0, _parent.OutputLength) [](start = 99, length = 42)
> 
> ToList().GetRange() may not be necessary. You can call Take() on the array

Resolved.